sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/MutationGroupEncoder.java

```diff
@@ -0,0 +1,649 @@
+    int word = i / 8;
+    int bit = 7 - i % 8;
+    bytes[word] |= 1 << bit;
+  }
+
+  private static boolean getBit(byte[] bytes, int i) {
+    int word = i / 8;
+    int bit = 7 - i % 8;
+    return (bytes[word] & 1 << (bit)) != 0;
+  }
+
+  private void encodeMutation(ByteArrayOutputStream bos, Mutation m) throws IOException {
+    Mutation.Op op = m.getOperation();
+    bos.write(op.ordinal());
+    if (op == Mutation.Op.DELETE) {
+      encodeDelete(bos, m);
+    } else {
+      encodeModification(bos, m);
+    }
+  }
+
+  private void encodeDelete(ByteArrayOutputStream bos, Mutation m) throws IOException {
+    List<String> allTables = schema.getTables();
+    String table = m.getTable().toLowerCase();
+    int tableIndex = allTables.indexOf(table);
+    VarInt.encode(tableIndex, bos);
+    ObjectOutput out = new ObjectOutputStream(bos);
+    out.writeObject(m.getKeySet());
+  }
+
+  private Mutation decodeDelete(ByteArrayInputStream bis)
+      throws IOException {
+    List<String> tables = schema.getTables();
+    int tableIndex = VarInt.decodeInt(bis);
+    String tableName = tables.get(tableIndex);
+
+    ObjectInputStream in = new ObjectInputStream(bis);
+    KeySet keySet;
+    try {
+      keySet = (KeySet) in.readObject();
+    } catch (ClassNotFoundException e) {
+      throw new RuntimeException(e);
+    }
+    return Mutation.delete(tableName, keySet);
+  }
+
+  private void encodeModification(ByteArrayOutputStream bos, Mutation m) throws IOException {
+    List<String> allTables = schema.getTables();
+    String tableName = m.getTable().toLowerCase();
+    int tableIndex = allTables.indexOf(tableName);
+    VarInt.encode(tableIndex, bos);
+    List<SpannerSchema.Column> columns = schema.getColumns(tableName);
+    checkArgument(columns != null, "Schema for table " + tableName + " not " + "found");
+    Map<String, Value> map = mutationAsMap(m);
+    // java.util.BitSet#toByteArray returns array of unpredictable length. Using byte arrays
+    // instead.
+    int bitsetSize = (columns.size() + 7) / 8;
+    byte[] exists = new byte[bitsetSize];
+    byte[] nulls = new byte[bitsetSize];
+    for (int i = 0; i < columns.size(); i++) {
+      String columnName = columns.get(i).getName();
+      boolean columnExists = map.containsKey(columnName);
+      boolean columnNull = columnExists && map.get(columnName).isNull();
+      if (columnExists) {
+        setBit(exists, i);
+      }
+      if (columnNull) {
+        setBit(nulls, i);
+        map.remove(columnName);
+      }
+    }
+    bos.write(exists);
+    bos.write(nulls);
+    for (int i = 0; i < columns.size(); i++) {
+      if (!getBit(exists, i) || getBit(nulls, i)) {
+        continue;
+      }
+      SpannerSchema.Column column = columns.get(i);
+      Value value = map.remove(column.getName());
+      encodeValue(bos, value);
+    }
+    checkArgument(map.isEmpty(),
+        "Columns " + map.keySet() + " were not defined in" + " table " + m.getTable());
```

Comment:
checkArgument messages should use format specifiers to avoid constructing the string concatenation every time even if the condition holds.

checkArgument(map.isEmpty(), "Columns %s were not defined in %s", ...)